File: 10_aula/Machine.cpp
#include <string>
#include <forward_list>
#include <iostream>


class Machine {
private:
  std::shared_ptr<MachineState> _state = std::make_shared<OpenOffState>( std::make_shared<Machine>());
public:
  void setState(std::shared_ptr<MachineState> state) { _state = state; }

  void power()  { _state->power();}
  void open()   { _state->open(); }
  void tick()   { _state->tick(); }
  void status() { std::cout << _state->status(); }
};

File: 10_aula/java/Machine.java
public class Machine {

    private MachineState _state = new NormalState(this);

    public void setState(MachineState state) {
        _state = state;
    }

    public void on() {
        System.out.print(_state + " --(on)-> ");
        _state.on();
        System.out.println(_state);
    }

    public void off() {
        System.out.print(_state + " --(off)-> ");
        _state.off();
        System.out.println(_state);
    }

    public void sound() {
        System.out.print(_state + " --(sound)-> ");
        _state.sound();
        System.out.println(_state);
    }
}